### PR TITLE
MOBT-689 (Pt 1): Move functionality out of the spot-extract CLI into a plugin

### PR DESCRIPTION
**Pull request by bayliffe**
Prior to adding new functionality to the spot-extract plugins it was necessary to simplify the CLI inline with the changes that are needed for implementing IMPROVER configurations using DAGRunner. These changes are:

- Move the logic from the CLI to a new plugin (SpotManipulation, hate the name and open to suggestions).
- Add unit tests to cover this logic which was previously only checked with acceptance tests.
- Make the `neighbour_finding_method_name` functionality a utility as it is used in several places. This includes the creation of a new unit test file for these spot-data utilities.
- Move the `check_grid_match` utility from the spotdata module to the metadata module where the related functionality lives.

Pt 2 of this work will add the functionality to extract a subset of spot sites from an existing spot-forecast cube using the spot-extract CLI / SpotManipulation plugin. This was large enough so I've stopped here to split it up, though Pt 2 should be much smaller.

Testing:
 - [x] Ran tests and they passed OK
 - [x] Added new tests for the new feature(s)

### CODE REVIEW

File: improver_tests/spotdata/test_SpotManipulation.py [hunk 1]
```diff
@@ -0,0 +1,320 @@
...
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+"""Unit tests for SpotManipulation class"""
```
Comment (brhooper) [inline improver_tests/spotdata/test_SpotManipulation.py:31]: Codecov has highlighted a couple of lines of code in `improver/spotdata/spot_manipulation.py` which aren't currently covered by tests.

L178: using the `realization_collapse` option

L200-201: using the `extract_percentiles` option with a realization cube

Can tests be added to cover these please?

**Review by brhooper — CHANGES_REQUESTED** [review]
Thanks @bayliffe, I've added a couple of comments.

**Review by brhooper — APPROVED** [review]
Thanks @bayliffe, I think this all looks fine now.

File: improver/spotdata/spot_manipulation.py [hunk 2]
```diff
@@ -0,0 +1,249 @@
+# -*- coding: utf-8 -*-
```
Comment (gavinevans) [inline improver/spotdata/spot_manipulation.py:1]: The new files in this PR: improver/spotdata/spot_manipulation.py, improver/spotdata/utilities.py, improver_tests/spotdata/test_SpotManipulation.py and improver_tests/spotdata/test_utilities.py now have a licence that conflicts with master following https://github.com/metoppv/improver/pull/2001, so you'll need to update these licences in your PR.

**Review by gavinevans — CHANGES_REQUESTED** [review]
Thanks @bayliffe 👍 

I've added some minor comments.

File: improver/spotdata/spot_manipulation.py [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+                        percentiles=extract_percentiles,
+                        fast_percentile_method=fast_percentile_method,
+                    )(result)
+                else:
+                    msg = (
+                        "Diagnostic cube is not a known probabilistic type. "
+                        "The {} percentile could not be extracted. Extracting "
+                        "data from the cube including any leading "
+                        "dimensions.".format(extract_percentiles)
+                    )
```
[the remark this answers is not in the document]
Comment (bayliffe) [reply improver/spotdata/spot_manipulation.py:212]: Moved.

File: improver/spotdata/utilities.py [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+
+"""Spot data utilities."""
+
+
+def neighbour_finding_method_name(land_constraint: bool, minimum_dz: bool) -> str:
```
[the remark this answers is not in the document]
Comment (bayliffe) [reply improver/spotdata/utilities.py:35]: Renamed.

File: improver/spotdata/spot_manipulation.py [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+                        "realization",
+                        percentiles=extract_percentiles,
+                        fast_percentile_method=fast_percentile_method,
+                    )(result)
+                else:
+                    msg = (
+                        "Diagnostic cube is not a known probabilistic type. "
+                        "The {} percentile could not be extracted. Extracting "
+                        "data from the cube including any leading "
+                        "dimensions.".format(extract_percentiles)
```
[the remark this answers is not in the document]
Comment (bayliffe) [reply improver/spotdata/spot_manipulation.py:211]: I've tweaked this to make it clearer (I hope).

File: improver/spotdata/spot_manipulation.py (hunk 2 above)
Comment (bayliffe) [reply improver/spotdata/spot_manipulation.py:1]: Done.

File: improver/spotdata/spot_manipulation.py [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+        apply_lapse_rate_correction: bool = False,
+        fixed_lapse_rate: Optional[float] = None,
+        land_constraint: bool = False,
+        similar_altitude: bool = False,
+        extract_percentiles: Optional[Union[float, List[float]]] = None,
+        ignore_ecc_bounds_exceedance: bool = False,
+        skip_ecc_bounds: bool = False,
+        new_title: Optional[str] = None,
+        suppress_warnings: bool = False,
+        realization_collapse: bool = False,
```
[the remark this answers is not in the document]
Comment (bayliffe) [reply improver/spotdata/spot_manipulation.py:76]: All of the other kwargs are passed through to either the get_neighbour_finding_method_name utility, ConvertProbabilitiesToPercentiles, or the spot-extract plugin. They are tested within the unit tests for those methods. I'm not sure what level of coverage to add to this wrapper without simply duplicating these tests at a higher level.

I've added a test to check that the self.neighbour_selection_method variable is set as expected when initialising the plugin as the actual setting of the variable is specific to this plugin even if the function has tests elsewhere.

**Review by gavinevans — APPROVED** [review]

**Pull request merged**